bio-bigfile: init 1.07

apraga opened the pull request:
Perl interface to manipulate BigWig and BigBeg files (for genomics)

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

vojta001 commented:
Result of `nixpkgs-review pr 186462` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>perl532Packages.BioBigFile</li>
    <li>perl534Packages.BioBigFile</li>
  </ul>
</details>

apraga commented:
@stigtsp @zakame Would it be possible to take a look at this ? Thanks :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1257

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      sha256 = "1455dwzpaq4hyhcqj3fpwgq5a39kp46qarfbr6ms6l2lz583r083";
+    };
+    # Use the ENSEMBL instructions, with some changes from
+    # https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/applications/science/biology/kent/default.nix#L70
```
```suggestion
```
that will be outdated real fast

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+    };
+    # Use the ENSEMBL instructions, with some changes from
+    # https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/applications/science/biology/kent/default.nix#L70
+    # We *cannot* use the package from nixpkgs (tests fails)
```
We could disable/fix the tests instead.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      ''
+      export MACHTYPE=$(uname -m)
+      export CFLAGS="-fPIC"
+      export MYSQLINC=`mysql_config --include | sed -e 's/^-I//g'`
+      export MYSQLLIBS=`mysql_config --libs`
```
```suggestion
      export MYSQLINC=$(mysql_config --include | sed -e 's/^-I//g')
      export MYSQLLIBS=$(mysql_config --libs)
```
Why are we not using nix variables to find those paths?

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+    # Use the ENSEMBL instructions, with some changes from
+    # https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/applications/science/biology/kent/default.nix#L70
+    # We *cannot* use the package from nixpkgs (tests fails)
+    buildPhase =
+      ''
+      export MACHTYPE=$(uname -m)
+      export CFLAGS="-fPIC"
```
```suggestion
    buildPhase = ''
      export MACHTYPE=$(uname -m)
      export CFLAGS="-fPIC"
```
I am not sure if uname -m is pure.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      echo 'CFLAGS="-fPIC"' > ../inc/localEnvironment.mk
+
+      # Build kent source
+      make clean && make
```
```suggestion
      make clean && make -j$NIX_BUILD_CORES
```
Why do we need to run make clean? Is the source not already clean?

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      # Build kent source
+      make clean && make
+      cd ../jkOwnLib
+      make clean && make
```
```suggestion
      make clean && make -j$NIX_BUILD_CORES
```

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      cp ../lib/x86_64/jkweb.a $NIX_BUILD_TOP/lib
+
+      # Modified: add include files
+      cp -r ../inc $NIX_BUILD_TOP/
+      '';
```
```suggestion
      cp -r ../inc $NIX_BUILD_TOP/
    '';
```

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+      make clean && make
+      cd ../jkOwnLib
+      make clean && make
+      mkdir -p $NIX_BUILD_TOP/bin/$MACH_TYPE
+      mkdir -p $NIX_BUILD_TOP/lib
+
+      cp ../lib/x86_64/jkOwnLib.a $NIX_BUILD_TOP/lib
+      cp ../lib/x86_64/jkweb.a $NIX_BUILD_TOP/lib
```
Referencing NIX_BUILD_TOP should be avoided as much as possible. Please try using relative paths.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+
+      # Only install lib and include. Lib must have the architecture
+      installPhase = ''
+      mkdir -p $out/lib/x86_64
+      cp $NIX_BUILD_TOP/lib/jkOwnLib.a $out/lib/x86_64/
+      cp $NIX_BUILD_TOP/lib/jkweb.a $out/lib/x86_64/
+      cp -r $NIX_BUILD_TOP/inc $out/
```
```suggestion
        mkdir -p $out/lib/x86_64
        cp $NIX_BUILD_TOP/lib/jkOwnLib.a $out/lib/x86_64/
        cp $NIX_BUILD_TOP/lib/jkweb.a $out/lib/x86_64/
        cp -r $NIX_BUILD_TOP/inc $out/
```
Same with NIX_BUILD_TOP here. Hardcoding x86_64 here is definitely impure and should not be done as it breaks builds on any other platform.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+  src = fetchurl {
+    url = "mirror://cpan/authors/id/L/LD/LDS/Bio-BigFile-${version}.tar.gz";
+    sha256 = "277b66ce8acbdd52399e2c5a0cf4e3bd5c74c12b94877cd383d0c4c97740d16d";
+  };
+  buildInputs = (with pkgs; with perlPackages; [ BioPerl IOString kent335 libpng libmysqlclient openssl ]);
```
```suggestion
  };

  buildInputs = [ BioPerl IOString kent335 libpng libmysqlclient openssl ];
```
pkgs should generally tried to be avided as much as possible and perlPackages should not be used inside of perlPackages.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+    sha256 = "277b66ce8acbdd52399e2c5a0cf4e3bd5c74c12b94877cd383d0c4c97740d16d";
+  };
+  buildInputs = (with pkgs; with perlPackages; [ BioPerl IOString kent335 libpng libmysqlclient openssl ]);
+  kentPath = builtins.toString kent335;
+
+  # Starting with GCC-11, compilation fails if -Wno-format-security
+  perlPreHook = ''
+  export KENT_SRC=$kentPath ;
+  export MACHTYPE=$(uname -m) ;
+  export CFLAGS=-Wno-format-security ;
+  '';
```
```suggestion

  perlPreHook = ''
    export KENT_SRC=${kent335}
    export MACHTYPE=$(uname -m)
    # Starting with GCC-11, compilation fails if -Wno-format-security
    export CFLAGS=-Wno-format-security
  '';

```

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,75 @@
+  '';
+  meta = with lib; {
+    description = "Manipulate Jim Kent's BigWig and BigBed index files for genomic features";
+    license = licenses.unfree;
```
Missing maintainer and homepage

apraga commented:
I've rewritten the pull request to use nix version of kent. This package only works with version 335 as written in the official documentation of VEP (https://www.ensembl.org/info/docs/tools/vep/script/vep_download.html), a future package that I intend to submit to nix.

So we need to merge 2 PRs before this one 
   - #206991 to build kent latest version
    - #223411 to package header files with kent

To build kent 335, I had to patch utils. There are not needed by Bio-BigFile but it looked "cleaner" to patch that to somehow edit the buildPhase.
Any advice is gladly appreciated.

apraga commented:
The requirements are now completed to merge this PR. @SuperSandro2000 could you take a look please ?

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,48 @@
+
+  perlPreHook = ''
+    export KENT_SRC=${kent}/
+    echo $KENT_SRC
+    # export MACHTYPE=$(uname -m)
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -67,15 +71,13 @@ stdenv.mkDerivation rec {
-    cp -r $HOME/inc/* $out/inc/
-
-    runHook postInstall
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    mkdir -p $out/inc
```
```suggestion
    mkdir -p $out/{bin,lib,inc}
```

SuperSandro2000 commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -67,15 +71,13 @@ stdenv.mkDerivation rec {
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    mkdir -p $out/inc
+    cp $NIX_BUILD_TOP/lib/jkOwnLib.a $out/lib
+    cp $NIX_BUILD_TOP/lib/jkweb.a $out/lib
+    cp $NIX_BUILD_TOP/bin/x86_64/* $out/bin
+    cp -r $NIX_BUILD_TOP/inc/* $out/inc/
```
Please use relative paths

SuperSandro2000 commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -67,15 +71,13 @@ stdenv.mkDerivation rec {
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    mkdir -p $out/inc
+    cp $NIX_BUILD_TOP/lib/jkOwnLib.a $out/lib
+    cp $NIX_BUILD_TOP/lib/jkweb.a $out/lib
```
```suggestion
    cp lib/jkOwnLib.a lib/jkweb.a $out/lib
```

SuperSandro2000 commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -52,13 +52,17 @@ stdenv.mkDerivation rec {
     chmod +x ./checkUmask.sh
     ./checkUmask.sh
 
+    mkdir -p $NIX_BUILD_TOP/lib
+    mkdir -p $NIX_BUILD_TOP/bin/x86_64
+    mkdir -p $NIX_BUILD_TOP/inc
```
same here, we can combine this and most likely use relative paths instead

SuperSandro2000 commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -24,9 +25,7 @@ stdenv.mkDerivation rec {
-  patchPhase = ''
-    runHook prePatch
-
+  prePatch = ''
```
```suggestion
  postPatch = ''
```

apraga commented:
Sorry for the late answer. I've taken into account your remarks. To make this PR compiles again, some changes have been introduced in the kent package:
    - postPatch was running in a loop
    - relative paths are used instead of NIX_BUILD_TOP
    - the upstream patch is only needed for kent 335

Both kent and Bio-BigFile compile.

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,55 @@
+      rev = "v${version}_base";
+      sha256 = "1455dwzpaq4hyhcqj3fpwgq5a39kp46qarfbr6ms6l2lz583r083";
+    };
+    patches = [ ./kent-utils.patch
+                (fetchpatch {
+                  url = "https://github.com/ucscGenomeBrowser/kent/commit/316e4fd40f53c96850128fd65097a42623d1e736.patch";
+                  sha256 = "sha256-wr3NP5qoSonKz1TLKtQyrTPErCOk2gC1RimcX0tE7cM=";
+                })
+              ];
+
```
```suggestion
    patches = [
      ./kent-utils.patch
      (fetchpatch {
        url = "https://github.com/ucscGenomeBrowser/kent/commit/316e4fd40f53c96850128fd65097a42623d1e736.patch";
        sha256 = "sha256-wr3NP5qoSonKz1TLKtQyrTPErCOk2gC1RimcX0tE7cM=";
      })
    ];
```

SuperSandro2000 commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,55 @@
+  perlPreHook = ''
+    echo "coucou"
+    export KENT_SRC=${kent335}/
+    echo ${kent335}
```
```suggestion
```

stigtsp commented:
@ofborg build perlPackages.BioBigFile kent

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+      sha256 = "1455dwzpaq4hyhcqj3fpwgq5a39kp46qarfbr6ms6l2lz583r083";
+    };
+    patches = [
+      ./kent-utils.patch
```
Could you add a comment to this patch? Is this a manual backport of a patch that was merged upstream?

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+  perlPreHook = ''
+    export KENT_SRC=${kent335}/
+    # Starting with GCC-11, compilation fails without -Wno-format-security
+    export CFLAGS=-Wno-format-security
```
I think the classic way to do that is with `hardeningDisable`, as explained [here](https://nixos.org/manual/nixpkgs/unstable/#sec-hardening-in-nixpkgs).

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+  buildInputs = [ BioPerl IOString libpng libmysqlclient openssl ];
+
+  perlPreHook = ''
+    export KENT_SRC=${kent335}/
```
You can do that with:

```nix
  KENT_SRC = kent.overrideAttrs(old: {
    # ...
  });
```

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+    sha256 = "277b66ce8acbdd52399e2c5a0cf4e3bd5c74c12b94877cd383d0c4c97740d16d";
+  };
+
+  buildInputs = [ BioPerl IOString libpng libmysqlclient openssl ];
```
Personal formatting preference, works better in future diffs.

```suggestion
  buildInputs = [
    BioPerl
    IOString
    libpng
    libmysqlclient
    openssl
  ];
```

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+{ lib, perl, buildPerlModule, fetchpatch, fetchurl, fetchFromGitHub, kent,
```
Same here regarding my personal, yet common formatting preference:

```suggestion
{ lib
, perl
, buildPerlModule
, fetchpatch
, fetchurl
, fetchFromGitHub
, kent
# ...
```

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,53 @@
+# 1. building utils fails (see patche below)
+# 2. we need headers (PR submitted for kent package)
+let
+  kent335 = kent.overrideAttrs (old: rec {
```
Is there a link to an upstream issue regarding Bio-BigFile not working with the latest `kent`? If so, put such links in comments.

apraga committed:
kent: enable patches if needed.

This is required by BioBigFile, which needs an older Kent version.

apraga committed:
perlPackages.BioBigfile: init at 1.07

Perl interface to manipulate BigWig and BigBeg files (for genomics)
This will be used to package VEP : uttps://www.ensembl.org/info/docs/tools/vep/script/vep_download.html

For that, only kent version 335 works at the moment.
For this commit, 2 commits for kent are needed
- e027a5d8368a053e231 : packaging kent header
- 2b2c6f4c287c6107acb : enabling patches in kent

apraga commented on pkgs/applications/science/biology/kent/default.nix:
```diff
@@ -24,16 +25,12 @@ stdenv.mkDerivation rec {
   buildInputs = [ libpng libuuid zlib bzip2 xz openssl curl libmysqlclient ];
 
-  patchPhase = ''
-    runHook prePatch
```
Thanks for pointing that out. I've cleaned up unnecessary changes in kent in the new version.

doronbehar replied:
> Thanks for pointing that out. I've cleaned up unnecessary changes in kent in the new version.

Your cleanup may have some sense, but if you'd let it go, you'd spare at least 1 rebuild from this PR, and you'd make the diffstat smaller.

apraga commented:
Thanks @doronbehar for your review. The code should be cleaner taking into account latest change in kent :
- only the possibility to add patches has been added
Your changes have been merged with a separate commit for kent.

I'm unsure why github is blocking the process as everything looks addressed to me.

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,66 @@
+    # The first patch is a custom fix  for linking error with zlib.
+    patches = [
+      ./kent-utils.patch
+      (fetchpatch {
```
I meant in [my previous comment](https://github.com/NixOS/nixpkgs/pull/186462/files#r1350704526), that a comment for this patch is missing.

apraga replied:
I've added a small comment in the new version. Is that enough ?

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,66 @@
+    IOString
+    libpng
+    libmysqlclient
+    openssl
```
Have you tried simply adding `zlib` to the inputs?

doronbehar replied:
If indeed adding `zlib` is fixing the issue, perhaps this change could be done to the main `kent/default.nix` expression.

apraga replied:
Unfortunately adding `zlib` is not enough due to some nested Makefile.

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,66 @@
+    openssl
+  ];
+
+  hardeningDisable = [ "format" ];
```
It would be nice to also give an explanation to why this is needed.

apraga replied:
A small comment has also been added, please check if it's okays.

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,66 @@
+  meta = with lib; {
+    homepage = "https://metacpan.org/dist/Bio-BigFile";
+    description = "Manipulate Jim Kent's BigWig and BigBed index files for genomic features";
+    license = licenses.unfree;
```
Because the license is `unfree`, ofborg doesn't build this package - making the CI status have less green checks.

apraga replied:
The license is actually GPL artistic 2, I've got confused with Kent's license.

doronbehar commented on pkgs/development/perl-modules/Bio-BigFile/default.nix:
```diff
@@ -0,0 +1,66 @@
+  # - official documentation: https://www.ensembl.org/info/docs/tools/vep/script/vep_download.html#bigfile
+  # - one of the developer's answer: https://github.com/Ensembl/ensembl-vep/issues/1412
+  # KENT_SRC is needed for building BioBigfil
+  KENT_SRC = kent.overrideAttrs (old: rec {
```
The attribute name `KENT_SRC` is not good in my opinion - `kent'` is more common for this case, the `_SRC` suffix is incorrect - this is not a source derivation.

apraga replied:
BioBigfile needs an environment variable named KENT_SRC to know where kent is installed. I cannot change that unfortunately.

doronbehar commented on pkgs/applications/science/biology/kent/kent-utils.patch:
```diff
@@ -0,0 +1,27 @@
+--- ./src/utils/bigWigCat/makefile.orig	2023-03-27 14:03:01.585896516 +0000
```
This patch is not used anywhere, and doesn't belong to this folder - should be deleted.

apraga replied:
As I've understand it, kent disable patching and this is needed for this older version. I haven't figured out another way of doing it.

doronbehar reviewed (changes requested):
Sorry for the slow replies - these are busy times!

apraga committed:
perlPackages.BioBigFile: Addressing reviewer change

apraga commented:
No worries, I am not the fastest either to address your change. Please check the new version if it is OK with you.

stigtsp reviewed (commented):
This looks good to me in general, with some nitpicks:

- The patch for `ucscGenomeBrowser/kent` should be vendored, since it's not merged into the upstream repo.
 - For the addition to `perlPackages.BioBigFile` please use that as prefix, like `perlPackages.BioBigFile: init at 1.07`
 - The commit "Addressing reviewer change" does not comply with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md). Please amend the relevant changes to the to other commits.

stigtsp commented:
Thx for your work on this 😀

apraga commented:
Thanks for your help merging this !
